src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs

```diff
@@ -99,84 +99,59 @@ public string RouteTemplate
-                {
-                    if (_logger.IsEnabled(LogLevel.Verbose))
-                    {
-                        _logger.WriteValues(CreateRouteAsyncValues(
-                            requestPath,
-                            context.RouteData.Values,
-                            matchedValues: false,
-                            matchedConstraints: false,
-                            handled: context.IsHandled));
-                    }
+            var values = _matcher.Match(requestPath);
 
-                    // If we got back a null value set, that means the URI did not match
-                    return;
-                }
+            if (values == null)
+            {
+                // If we got back a null value set, that means the URI did not match
 
-                var oldRouteData = context.RouteData;
+                _logger.LogVerbose(
+                        $"Request did not match the route with name '{Name}' and template '{RouteTemplate}'.");
 
-                var newRouteData = new RouteData(oldRouteData);
-                MergeValues(newRouteData.DataTokens, _dataTokens);
-                newRouteData.Routers.Add(_target);
-                MergeValues(newRouteData.Values, values);
+                return;
+            }
 
-                if (!RouteConstraintMatcher.Match(
-                    Constraints,
-                    newRouteData.Values,
-                    context.HttpContext,
-                    this,
-                    RouteDirection.IncomingRequest,
-                    _constraintLogger))
-                {
-                    if (_logger.IsEnabled(LogLevel.Verbose))
-                    {
-                        _logger.WriteValues(CreateRouteAsyncValues(
-                            requestPath,
-                            newRouteData.Values,
-                            matchedValues: true,
-                            matchedConstraints: false,
-                            handled: context.IsHandled));
-                    }
+            var oldRouteData = context.RouteData;
 
-                    return;
-                }
+            var newRouteData = new RouteData(oldRouteData);
+            MergeValues(newRouteData.DataTokens, _dataTokens);
+            newRouteData.Routers.Add(_target);
+            MergeValues(newRouteData.Values, values);
 
-                try
-                {
-                    context.RouteData = newRouteData;
+            if (!RouteConstraintMatcher.Match(
+                Constraints,
+                newRouteData.Values,
+                context.HttpContext,
+                this,
+                RouteDirection.IncomingRequest,
+                _constraintLogger))
+            {
+                return;
+            }
 
-                    await _target.RouteAsync(context);
+            _logger.LogInformation(
+                $"Request successfully matched the route with name '{Name}' and template '{RouteTemplate}'.");
```

Comment:
:+1: I think we should, for example in the case where route matching succeeds but action selection fails, route values would be useful in diagnosis.